### PR TITLE
Adding padded layout 'layout_padded_general' 

### PR DESCRIPTION
**Pull request by mfoerste4**
This is a different approach / followup PR of https://github.com/rapidsai/raft/pull/663 for issue https://github.com/rapidsai/raft/issues/497.

I implemented a `layout_padded_general` within raft to statically enforce padding on mdpsan accesses.
* The layout has template parameters for `ValueType`, `StorageOrder `(default `row_major_t`), and `ByteAlignment `(default 128)
* in order to *not* require changes upstream I skipped `submdspan `functionality right now. I have a branch on a mdspan fork where I tested this though (https://github.com/mfoerste4/mdspan/tree/layout_padded).

### CODE REVIEW

**Comment by mfoerste4** [comment]
@achirkin, please have a look at this. It is now possible to retrieve the layout / padding width at compile time which makes it possible to target optimized kernel code correctly.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (achirkin) [inline cpp/include/raft/detail/layout_padded_general.hpp:65]: A nitpick: perhaps, we can use `raft::ceildiv` for readability?

File: cpp/include/raft/matrix/detail/linewise_op.cuh
Comment (achirkin) [inline cpp/include/raft/matrix/detail/linewise_op.cuh:211]: Out of curiosity: why ones and not zeroes? :)

Comment (mfoerste4) [reply cpp/include/raft/matrix/detail/linewise_op.cuh:211]: I had division ops in mind and did not want to risk any division by zero. I was unsure whether this might cause issues with tools like valgrind?

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (achirkin) [reply cpp/include/raft/detail/layout_padded_general.hpp:77]: Sounds good. Maybe you'd also want to constrain it here to always be power-of-two (I've noticed you used Pow2<..> utils somewhere)? That would still cover our hardware-inspired use cases.

File: cpp/include/raft/matrix/detail/linewise_op.cuh
Comment (achirkin) [reply cpp/include/raft/matrix/detail/linewise_op.cuh:211]: Hmm, good idea, I guess. I don't think it will cost anything anyway.

**Comment by mfoerste4** [comment]
> Looks good! I wonder if we can further simplify/optimize some parts of the `linewiseOp` for padded data?..

The current implementation basically consists of the old main kernel which was running the aligned data portion. I don't think we can simplify it further without loosing performance. Regarding optimizations - we could skip the operation on the padded portion of data, but I guess that would only have very limited effect on larger datasets.

**Comment by mfoerste4** [comment]
@achirkin , what are the next steps here to proceed?

File: cpp/include/raft/matrix/detail/linewise_op.cuh
Comment (achirkin) [inline cpp/include/raft/matrix/detail/linewise_op.cuh:33]: Perhaps, this, parameterized by the padding size, should go to `raft/core/mdarray.hpp`?..

File: cpp/include/raft/matrix/detail/linewise_op.cuh
Comment (achirkin) [inline cpp/include/raft/matrix/detail/linewise_op.cuh:36]: The same here (so that the api in `matrix.cuh` doesn't show anything in `detail` namespace).

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (achirkin) [inline cpp/include/raft/detail/layout_padded_general.hpp:77]: @mhoemmen, sorry to call you out again, could you have a quick look at our attempt to define the padded layout - whether it is properly in the spirit of the mdspan spec?

**Review by achirkin — CHANGES_REQUESTED** [review]
Overall, this looks good to me.

The only thing I'm not sure is whether we'd want to define the padded layout synonyms for public use _with or without the template parameter `padding size`_. On the one hand, the parameterized version seems to be more logical for whatever use case. On the other hand, any application I could imagine, it only matters to have the specific padding of 128 bytes for the coalesced memory access. Or,  maybe, we should have both?

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mhoemmen) [reply cpp/include/raft/detail/layout_padded_general.hpp:77]: @achirkin I'm happy to help out! : - )

My main issue with this design is that the layout mapping cannot possibly be responsible for alignment.  The layout can do padding to a particular size, but it can't control whether the pointer is aligned (so that you can do aligned SIMD loads, etc.).  This is because the layout never sees the pointer; it's the accessor that sees the pointer.

Would you instead consider the following design?

* The layout mapping 
    * pads each row (if row-major, or column, if column-major) to a given number of _elements_ (not a number of bytes); and
    * encodes the padding as a `size_type` or `integral_constant<size_type, Value>` template parameter.
* The accessor
    * encodes the alignment as a `size_t` or `integral_constant<size_t, Value>`;
    * defines the `pointer` type with the appropriate attributes to tell the compiler that it is aligned; and
    * has a different `offset_policy` (e.g., `default_accessor`) to express that offset access may not be aligned.

A type alias and possibly also a helper create-an-mdspan function would then use the `value_type` to compute the padding and alignment at compile time.  For example:
```c++
// Specialize this for different value types as needed.
// The value can be size_t to simplify specialization;
// layout_padded_general can convert it to size_type.
template<class ValueType>
struct padding {
    static constexpr size_t value = 128 / sizeof(ValueType);
};

template<class ElementType, class Extents, StorageOrderType order>
using aligned_mdspan = mdspan<ElementType, Extents, layout_padded_general<order, padding_v<remove_cv_t<remove_reference_t<ElementType>>>>, aligned_accessor<ElementType, alignment_v<remove_cv_t<remove_reference_t<ElementType>>>>>

template<class ElementType, class Extents, StorageOrderType order>
aligned_mdspan<ElementType, Extents, order>
make_aligned_mdspan(ElementType* input_pointer, Extents e, StorageOrderType /* order */)
{
    using value_type = remove_cv_t<remove_reference_t<ElementType>>;
    using pointer = typename aligned_accessor<ElementType, alignment_v<value_type>>::pointer;

    // Any run-time check for alignment would happen here.
    // Taking the alignment value as integral_constant permits interesting checks,
    // like wrapping ElementType* in a proxy pointer type
    // that encodes alignment as a non-type template parameter.
    pointer aligned_pointer = 
        assert_aligned(input_pointer, integral_constant<size_t, alignment_v<value_type>>{});
    using mapping = layout_padded_general<order, padding_v<value_type>>::template mapping<Extents>;
    return {aligned_pointer, mapping{e}};
>, 
```

**Review by mhoemmen — COMMENTED** [review]
Thanks for soliciting my review!  I have a few comments on the design.

File: cpp/include/raft/core/cudart_utils.hpp
Comment (achirkin) [reply cpp/include/raft/core/cudart_utils.hpp:81]: Good idea but NB we cannot remove `isPo2` because it's used in some places in raft and cuml and potentially in other projects. We just move here a few functions from `*.cuh` to `*.hpp` to make them available in the host-only code.

File: cpp/include/raft/core/mdarray.hpp
Comment (mhoemmen) [inline cpp/include/raft/core/mdarray.hpp:96]: [INFORMATIVE NOTE]

I've been thinking about how to express this idea.  The issue is that `std::assume_aligned` (C++20) or equivalent compiler built-ins (e.g., GCC's `__builtin_assume_aligned`) don't affect the type.  My current thinking is that `aligned_accessor::access` should do `return (std::assume_aligned(p))[i];`.  Please see my comment below on `aligned_accessor::access`.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mhoemmen) [inline cpp/include/raft/detail/layout_padded_general.hpp:158]: [This is not strictly required for correctness, but I'm guessing you won't get any benefit out of `aligned_accessor` unless you do something here.]

This would be the point at which I make assertions about alignment.  For example:
```c++
constexpr reference access(data_handle_type p, size_t i) const noexcept {
#if defined(__cpp_lib_assume_aligned)
    return (std::assume_aligned<alignment>(p))[i];
#elif defined(__GNUC__)
    return (__builtin_assume_aligned(p, alignment))[i];
#else
    return p[i];
#endif
}
```

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mfoerste4) [reply cpp/include/raft/detail/layout_padded_general.hpp]: Thanks for the hint - I just merged with the new tip and added the type alias.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mfoerste4) [reply cpp/include/raft/detail/layout_padded_general.hpp]: Yes, for now I was just switching to the right accessor class in order to allow for both static and runtime checks later. Thanks for example, I'll work that in.

File: cpp/include/raft/core/mdarray.hpp
Comment (mhoemmen) [inline cpp/include/raft/core/mdarray.hpp:77]: Would you consider getting this default index type from some common source, to avoid code duplication and possible errors between `matrix_extent` and (say) `vector_extent`?

File: cpp/include/raft/core/mdarray.hpp
Comment (mhoemmen) [inline cpp/include/raft/core/mdarray.hpp:113]: Given that `uint32_t` is the default, would you consider just relying on the default here, in order to avoid confusion and possible errors?

File: cpp/include/raft/core/mdarray.hpp
Comment (mfoerste4) [reply cpp/include/raft/core/mdarray.hpp:77]: There were several issues after merging with the introduced IndexType, I did not mean to change that here.

File: cpp/include/raft/core/mdarray.hpp
Comment (mfoerste4) [reply cpp/include/raft/core/mdarray.hpp]: @mhoemmen , thank you very much for your input. I have tried to apply all of your suggestions and also modified the matrix `LinewiseOp `operations to utilize the `aligned_mdspan`.

**Comment by mfoerste4** [comment]
@mhoemmen , thank you very much for your input. I have tried to apply all of your suggestions and also modified the matrix `LinewiseOp `operations to utilize the `aligned_mdspan`.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mhoemmen) [reply cpp/include/raft/detail/layout_padded_general.hpp]: @mfoerste4 Some compilers let one express alignment by annotating the pointer; e.g., `double* __attribute__(align_value( alignment ))` -- I haven't experimented with that, but I personally would prefer using the C++ Standard way (`std::assume_aligned`, as above) to the list of `ifdef`s.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (achirkin) [inline cpp/include/raft/detail/layout_padded_general.hpp:32]: Just a neatpick: could you check that raft's clang-tidy does not complain about the naming conventions for the arguments? I believe, the `Alignment` should be lower-case.

(Usually I set up vscode to check this: `clang-tidy -p ${fileWorkspaceFolder}/cpp/build ${file} -extra-arg=-std=c++17`)

**Review by achirkin — COMMENTED** [review]
Thanks, looks good to me! Just a small suggestion to check if clang-tidy does not complain too much.

BTW, did you have a chance to see how `linewiseOpSpan` compares in terms of speed to `linewiseOp`?

**Review by mhoemmen — COMMENTED** [review]
I think somehow my review never made it through, yikes!

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (mfoerste4) [reply cpp/include/raft/detail/layout_padded_general.hpp:32]: I have adjusted the specific parameter you mentioned. Running clang-tidy gives me several more warnings like ` warning: invalid case style for constexpr variable 'storage_order' [readability-identifier-naming]`, but the suggested naming `kStorageOrder ` does seemingly not match our conventions looking at other places in the code. Is my call missing the correct `.clang-tidy` options?

**Comment by mfoerste4** [comment]
> Thanks, looks good to me! Just a small suggestion to check if clang-tidy does not complain too much.
> 
> BTW, did you have a chance to see how `linewiseOpSpan` compares in terms of speed to `linewiseOp`?
I did some benchmarks on a V100. For most matrix shapes the span-implementation is as fast as the existing implementation with aligned data. It suffers significantly if the line-length is comparably small w.r.t. the amount of padding.

File: cpp/include/raft/detail/layout_padded_general.hpp
Comment (achirkin) [reply cpp/include/raft/detail/layout_padded_general.hpp]: That's correct `.clang-tidy` and I don't know what's the better option :) I usually choose to blindly follow clang-tidy suggestion, but I wouldn't ask you to do the same

**Review by achirkin — APPROVED** [review]
This looks good to me. Looking forward to try this in https://github.com/rapidsai/raft/issues/711 (1)!

**Comment by mhoemmen** [comment]
@mfoerste4 Just to follow up: my mdspan PR https://github.com/kokkos/mdspan/pull/176 adds an `aligned_accessor` example.

File: cpp/include/raft/core/mdarray.hpp
Comment (mhoemmen) [reply cpp/include/raft/core/mdarray.hpp]: For my latest thoughts on `assume_aligned` etc., please see my `aligned_accessor` PR https://github.com/kokkos/mdspan/pull/176 .

**Comment by mhoemmen** [comment]
@mfoerste4 I've been working on a different approach.  The idea is that a padded layout mapping has the same mapping function as a submdspan of a `layout_left` (for column major, or `layout_right` for row-major) mdspan, with suitably padded extents.  This would make `padded_layout::mapping` just a wrapper around `layout_left::mapping` (or `layout_right::mapping`).

This approach has the following advantages.

1. The only nontrivial code is computing the padded extents.  The [`pad_extents` function here here](https://godbolt.org/z/sv7daYGf1) shows how to do that.  Everything else is just a wrapper.  That reduces the testing burden.
2. If the original extents are all known at compile time, the padded extents can also be stored as compile-time values.  This makes storing and passing around padded-layout mdspans cheaper, and can improve compiler optimizations.

**Comment by mhoemmen** [comment]
@mfoerste4 Here is my proposed `layout_padded` with some simple tests: https://godbolt.org/z/odhbzx5eq

**Comment by mfoerste4** [comment]
> @mfoerste4 Here is my proposed `layout_padded` with some simple tests: https://godbolt.org/z/odhbzx5eq
> 
> My extents padding might be more generous than needed -- I think only the first or last extent needs padding, not the middle ones as well -- but that should be easy to fix.

@mhoemmen , I like the idea of compile time 'override' of the extents in your approach. It also feels more natural to base the padded layout on the left/right layout instead of utilizing the more general `strided_layout `with some implicit restrictions regarding the strides. 

Like you already mentioned the padding should only occur on first/last extent w.r.t. the layout. The middle ones don't need extra padding as the strides are implicitly aligned as multiples of the first/last extent respectively.

Regarding next steps - are you planning to merge this idea to the mdspan repo in the near future? Short term I could update the layout class of this PR, but it would be great to have it in the officical repo to avoid future merge issues.

**Comment by mhoemmen** [comment]
@mfoerste4 Thanks for your response and the feedback!

> Like you already mentioned the padding should only occur on first/last extent w.r.t. the layout. 

I'll fix that -- thanks!

> Regarding next steps - are you planning to merge this idea to the mdspan repo in the near future?

Yes -- I'm planning to make a PR in that repo.

Regarding `aligned_accessor`, Christian, Damien, and I have been debating some fine details of my PR https://github.com/kokkos/mdspan/pull/176 .  I think it's close to merging.  (mdspan PRs need approving reviews to merge.)  `layout_padded` doesn't depend on that, but in order to have overaligned mdspan, we will need to combine `aligned_accessor` and `layout_padded` and have examples and tests for their combination.

Btw, I find it a bit awkward that we say "`something_accessor`" but "`layout_something`."  The original convention was to put the specific "something" last: `accessor_basic`.  That got renamed to `default_accessor`, but the renaming didn't touch the layouts; I don't know why.  Thus, I think we should feel free to write "`padded_layout`" if we like, or "`layout_padded`," whichever we prefer.  Which would you prefer?

**Comment by mhoemmen** [comment]
@mfoerste4 Here is my `layout_padded` example's PR: https://github.com/kokkos/mdspan/pull/180

**Comment by mfoerste4** [comment]
> Which would you prefer?
I would prefer `layout_something`, as it is consistent with the other layouts we probably don't want to switch anymore. 

> @mfoerste4 Here is my `layout_padded` example's PR: [kokkos/mdspan#180](https://github.com/kokkos/mdspan/pull/180)
I initially thought you were planning to add the `layer_padded ` as a reusable header part in the [p0009](https://github.com/kokkos/mdspan/tree/stable/include/experimental/__p0009_bits). Having it available as an example is great, but would still require copy/pasting class code for consumers (e.g. raft). Having it generally available would also allow for proper treatment in the `submdspan` going forward. What do you think?

**Comment by mhoemmen** [comment]
@mfoerste4 wrote:
> I initially thought you were planning to add the `layout_padded` as a reusable header part in the [p0009](https://github.com/kokkos/mdspan/tree/stable/include/experimental/__p0009_bits). Having it available as an example is great, but would still require copy/pasting class code for consumers (e.g. raft). Having it generally available would also allow for proper treatment in the submdspan going forward. What do you think?

It's definitely the plan to make `layout_padded` and `aligned_accessor` reusable headers in the reference mdspan implementation.  However, I'm still talking with the other reference mdspan developers about how to do that (e.g., header path, namespace name, testing requirements).  If you have any suggestions, we would welcome them! : - ) 

For now, I just wanted to make the code available in some way, as quickly as possible.  I'll be happy to add `layout_padded` and `aligned_accessor` to RAFT itself for now, if you like.  Once mdspan gets them as reusable headers, we'll be able to remove those implementations.  Do you think that would be a good idea?

**Comment by mfoerste4** [comment]
>  I'll be happy to add layout_padded and aligned_accessor to RAFT itself for now, if you like. Once mdspan gets them as reusable headers, we'll be able to remove those implementations. Do you think that would be a good idea?

I can switch the `layout_padded_general` in this PR to your `layout_padded ` based on your example and modify the test and benchmark code for now. That way we hopefully have not much to change once it is generally available in mdspan. Additionally, we might get some valuable feedback from other people actually using it in the open field.

**Comment by mfoerste4** [comment]
@mhoemmen,  I noticed quite some activity on the layout_padded branch https://github.com/kokkos/mdspan/pull/180. I suppose we can use this implementation for now and move to the mdspan version once it is fully integrated.

**Comment by mhoemmen** [comment]
@mfoerste4 I'm almost done!  I should be finishing up the last few changes today.  It's a different design, based on discussions last week with Christian Trott about what we would like to try to standardize.  It should also be able to replace the "BLAS general" layouts in P1673.

**Comment by mhoemmen** [comment]
@mfoerste4 Please see updated PR here: https://github.com/kokkos/mdspan/pull/180

**Comment by mfoerste4** [comment]
I have updated header versions of `layout_{left_right}_padded` as well as `aligned_accessor` based on the examples in https://github.com/kokkos/mdspan/pull/180 within the thirdparty mdspan include directory (including disclaimer).
A wrapper layout and make utility is located in `cpp/include/raft/core/mdspan.hpp`. 
@cjnolet, could you assign a code-owner for final review?

File: cpp/include/raft/thirdparty/mdspan/include/experimental/mdspan
Comment (cjnolet) [inline cpp/include/raft/thirdparty/mdspan/include/experimental/mdspan:47]: Is this change already integrated into Kokkos upstream? Just making sure we aren't diverging so when we overwrite these files with upstream changes we aren't finding a bunch of unexpected breaking changes.

File: cpp/include/raft/core/mdspan.hpp
Comment (cjnolet) [inline cpp/include/raft/core/mdspan.hpp:40]: Can we stick this in raft/core/detail/mdspan.hpp just to keep the public API clean?

File: cpp/test/mdarray.cu
Comment (cjnolet) [inline cpp/test/mdarray.cu:768]: So for a padded mdarray, does this require that we use the accessor directly and can't, for example, just use the underlying data_handle/pointer? Thinking about the case where we have a matrix and want it padded so we call down to something like cublas, which would be expecting a raw pointer.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Thanks for working on this implementation, Malte! I've done a light review after catching myself up on the conversation in the thread. It looks like this PR could be brought up to date with branch-22.12 as well. Looking forward to having this feature in RAFT!

File: cpp/include/raft/thirdparty/mdspan/include/experimental/mdspan
Comment (mhoemmen) [reply cpp/include/raft/thirdparty/mdspan/include/experimental/mdspan:47]: @cjnolet I haven't had a chance to look at these changes much.  The reference mdspan implementation (it's not correct to call that "Kokkos"; it just happens to live in that repository but it is a separate project) does not yet have this header file as a reusable component.  `aligned_accessor` does exist there as an example.  However, the project leads are thinking about how to provide "supported extensions" like this as reusable components.

File: cpp/include/raft/core/mdspan.hpp
Comment (mhoemmen) [reply cpp/include/raft/core/mdspan.hpp:40]: P2642 has moved away from using tag types to express storage order.  It provides separate types `layout_left_padded` and `layout_right_padded`, for consistency with the Standard layouts `layout_left` and `layout_right`.

File: cpp/test/mdarray.cu
Comment (mhoemmen) [inline cpp/test/mdarray.cu:559]: Does Thrust have a way to do `cartesian_product(iota(0, rows), iota(0, cols))` iteration?  It's painful to have to go back and forth between 1-D and 2-D indices like this.

File: cpp/test/mdarray.cu
Comment (mhoemmen) [reply cpp/test/mdarray.cu:768]: @cjnolet This might relate to how mdarray's design has changed.  The latest version of the mdarray proposal has a way to get an mdspan viewing the mdarray's elements, with a given accessor.

mdarray always has a container underneath, so you should always be able to use the mdarray's pointer directly.  On the other hand, a much nicer way to access the mdarray's elements on device would be to get an mdspan viewing the elements.

File: cpp/test/mdarray.cu
Comment (mfoerste4) [reply cpp/test/mdarray.cu:559]: I could not find anything like that. Of course, one could define iterators for each dimension and zip them together, but that would just shift the stride-logic to another piece of manual code.

File: cpp/include/raft/matrix/matrix.cuh
Comment (mfoerste4) [reply cpp/include/raft/matrix/matrix.cuh]: I have aligned the interface as well as the template names with the given host/device pattern.

File: cpp/include/raft/core/mdspan.hpp
Comment (mfoerste4) [reply cpp/include/raft/core/mdspan.hpp:35]: I have removed the enum and provided two separate layouts.

**Comment by mfoerste4** [comment]
@cjnolet , thanks for reviewing. I re-based the branch to 22.12 and tried to align naming of template classes and API as close to the existing pattern as possible.

File: cpp/test/mdarray.cu
Comment (mfoerste4) [reply cpp/test/mdarray.cu:581]: I added https://github.com/rapidsai/raft/issues/922 in order to track.

**Review by cjnolet — DISMISSED** [review]

**Comment by mhoemmen** [comment]
@cjnolet @mfoerste4 I've been completely overwhelmed with my current project and haven't had time to review this.  Please don't feel like you have to wait on me, though I appreciate that you asked me for feedback!

**Review by cjnolet — APPROVED** [review]

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**
